### PR TITLE
[SPARK-37085][PYTHON][SQL] Add list/tuple overloads to array, struct, create_map, map_concat

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This PR adds overloads to the following `pyspark.sql.functions`:

- `array`
- `struct`
- `create_map`
- `map_concat`

to support calls with a single `list` or `tuple` argument, i.e.

```python
array(["foo", "bar"])
```

### Why are the changes needed?

These calls are supported by the current implementation, but don't type check.

### Does this PR introduce _any_ user-facing change?

Type checker only, as described above.

### How was this patch tested?

Existing tests and manual tests (to be added in SPARK-36989)

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -3514,7 +3538,19 @@ def map_from_arrays(col1: "ColumnOrName", col2: "ColumnOrName") -> Column:
...
+    ...
+
+
+@overload
+def array(__cols: Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]) -> Column:
+    ...
+
+
+def array(
+    *cols: Union["ColumnOrName", Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]]
```
Comment (xinrong-databricks) [inline python/pyspark/sql/functions.py:3552]: Out of curiosity would flattening unions of unions explicitly be better?

Comment (zero323) [reply python/pyspark/sql/functions.py:3552]: In general, nested `Unions` are flattened automatically:

```python
>>>  Union["ColumnOrName", Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]]
typing.Union[ForwardRef('ColumnOrName'), typing.List[ForwardRef('ColumnOrName')], typing.Tuple[ForwardRef('ColumnOrName'), ...
```

Personally, I find structure that clearly groups similar categories useful (for similar problem see https://github.com/apache/spark/pull/34671#discussion_r754532969), but I am not very attached to this idea.

Comment (xinrong-databricks) [reply python/pyspark/sql/functions.py:3552]: Makes sense. Thanks!

**Comment by HyukjinKwon** [comment]
@itholic can you take a look please?

**Comment by itholic** [comment]
Could you rebase this branch to master ??

mypy annotation tests seems not to be passed with current status

**Comment by zero323** [comment]
> Could you rebase this branch to master ??
> 
> mypy annotation tests seems not to be passed with current status

@itholic Done. It seems to pass everything in CI. Did you have any particular issues in mind (I've seen some weird local issues, until I cleared mypy cache after the upgrade).

**Comment by itholic** [comment]
@zero323 hmm.. I didn't see anything particularly strange except for a few issues caused by different mypy version. It seems to be working fine now.

**Comment by itholic** [comment]
Oh, `python/pyspark/pandas/indexes/base.py:48: error: Module "pandas._libs" has no attribute "lib"  [attr-defined]` is started to fail. Do you have same issue on your local ??

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1670,7 +1671,19 @@ def expr(str: str) -> Column:
...
 
 
+@overload
 def struct(*cols: "ColumnOrName") -> Column:
+    ...
+
+
+@overload
+def struct(__cols: Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]) -> Column:
+    ...
```
Comment (itholic) [inline python/pyspark/sql/functions.py:1681]: Maybe seems like `List["ColumnOrName_"]` just works fine.

Could you briefly explain why we need to use `Union` the `List` and `Tuple` ??

And could I ask if what is the `...` mean in `["ColumnOrName_", ...]` ??

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -1670,7 +1671,19 @@ def expr(str: str) -> Column:
...
+
+
+def struct(
+    *cols: Union["ColumnOrName", Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]]
+) -> Column:
```
Comment (itholic) [inline python/pyspark/sql/functions.py:1686]: ditto ?

**Comment by zero323** [comment]
> @zero323 hmm.. I didn't see anything particularly strange except for a few issues caused by different mypy version. It seems to be working fine now.

I don't think I've seen this one. In my local env I primarily see some errors caused by output mismatch in data tests, dynamically defined methods and stuff covered by https://github.com/apache/spark/pull/34946/, but all of these are non-deterministic and rare, so I don't even have good place to start debugging and asking questions.

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (zero323) [reply python/pyspark/sql/functions.py:1681]: We still want to support calls like

```python
struct(("foo", "bar"))
```

which shouldn't be accepted without `Tuple` (or some supertype).

If you try

```python
diff --git a/python/pyspark/sql/functions.py b/python/pyspark/sql/functions.py
index 006d10c9fc..caf17a84b3 100644
--- a/python/pyspark/sql/functions.py
+++ b/python/pyspark/sql/functions.py
@@ -1677,13 +1677,11 @@ def struct(*cols: "ColumnOrName") -> Column:
 
 
 @overload
-def struct(__cols: Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]) -> Column:
+def struct(__cols: Union[List["ColumnOrName_"]]) -> Column:
     ...
 
 
-def struct(
-    *cols: Union["ColumnOrName", Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]]
-) -> Column:
+def struct(*cols: Union["ColumnOrName", Union[List["ColumnOrName_"]]]) -> Column:
     """Creates a new struct column.
 
     .. versionadded:: 1.4.0
diff --git a/python/pyspark/sql/tests/typing/test_functions.yml b/python/pyspark/sql/tests/typing/test_functions.yml
index efb3293472..fdf2303890 100644
--- a/python/pyspark/sql/tests/typing/test_functions.yml
+++ b/python/pyspark/sql/tests/typing/test_functions.yml
@@ -44,6 +44,7 @@
     struct([col("foo"), col("bar")])
     struct("foo", "bar")
     struct(["foo", "bar"])
+    struct(("foo", "bar"))
 
     array([col("foo")], [col("bar")])
     create_map([col("foo")], [col("bar")])
```

you should see error in data tests.



> And could I ask if what is the `...` mean in `["ColumnOrName_", ...]` ??

`Tuples` are typed like product types, so `Tuple[ColumnOrName_]` matches tuple with exactly one column or str element. In contrast `Tuple["ColumnOrName_", ...]` matches tuples of arbitrary size, as long as all elements are columns or strings (there is [mypy doc section](https://mypy.readthedocs.io/en/stable/kinds_of_types.html?highlight=namedtuple#tuple-types) that discusses this further).

Comment (zero323) [reply python/pyspark/sql/functions.py:1681]: We still want to support (it was common user request in the past) calls like

```python
struct(("foo", "bar"))
```

which shouldn't be accepted without `Tuple` (or some supertype).

If you try

```python
diff --git a/python/pyspark/sql/functions.py b/python/pyspark/sql/functions.py
index 006d10c9fc..caf17a84b3 100644
--- a/python/pyspark/sql/functions.py
+++ b/python/pyspark/sql/functions.py
@@ -1677,13 +1677,11 @@ def struct(*cols: "ColumnOrName") -> Column:
 
 
 @overload
-def struct(__cols: Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]) -> Column:
+def struct(__cols: Union[List["ColumnOrName_"]]) -> Column:
     ...
 
 
-def struct(
-    *cols: Union["ColumnOrName", Union[List["ColumnOrName_"], Tuple["ColumnOrName_", ...]]]
-) -> Column:
+def struct(*cols: Union["ColumnOrName", Union[List["ColumnOrName_"]]]) -> Column:
     """Creates a new struct column.
 
     .. versionadded:: 1.4.0
diff --git a/python/pyspark/sql/tests/typing/test_functions.yml b/python/pyspark/sql/tests/typing/test_functions.yml
index efb3293472..f5f2f13c4f 100644
--- a/python/pyspark/sql/tests/typing/test_functions.yml
+++ b/python/pyspark/sql/tests/typing/test_functions.yml
@@ -66,6 +66,8 @@
     create_map(col_objs)
     map_concat(col_objs)
 
+    struct(("foo", "bar"))
+
   out: |
     main:29: error: No overload variant of "array" matches argument types "List[Column]", "List[Column]"  [call-overload]
     main:29: note: Possible overload variant:

```

you should see error in data tests

```
___________________________ varargFunctionsOverloads ___________________________
/path/to/spark/python/pyspark/sql/tests/typing/test_functions.yml:19: 
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Actual:
E     main:50: error: No overload variant of "struct" matches argument type "Tuple[str, str]"  [call-overload] (diff)
E     main:50: note: Possible overload variants:    (diff)
E     main:50: note:     def struct(*cols: Union[Column, str]) -> Column (diff)
E     main:50: note:     def [ColumnOrName_] struct(List[ColumnOrName_]) -> Column (diff)
E   Expected:
E     (empty)
=========================== short test summary info ============================
````


> And could I ask if what is the `...` mean in `["ColumnOrName_", ...]` ??

`Tuples` are typed like product types, so `Tuple[ColumnOrName_]` matches tuple with exactly one column or str element. In contrast `Tuple["ColumnOrName_", ...]` matches tuples of arbitrary size, as long as all elements are columns or strings (there is [mypy doc section](https://mypy.readthedocs.io/en/stable/kinds_of_types.html?highlight=namedtuple#tuple-types) that discusses this syntax further).

Comment (itholic) [reply python/pyspark/sql/functions.py:1681]: Got it. Thanks for the comment!! 🙏

**Review by itholic — APPROVED** [review]
It seems to be the best for now, until mypy itself supports looser typing conventions.

**Review by HyukjinKwon — APPROVED** [review]
Haven't taken a close look but I am fine with this.

**Comment by zero323** [comment]
Merged into master.

Thanks everyone!

**Comment by zero323** [comment]
> t seems to be the best for now, until mypy itself supports looser typing conventions.

As @ueshin pointed out, using `Sequence` might provide more general approach (and allow us to forget about variance), but there are two problems:

- It is high level interface that potentially covers more than lists and tuples. So this would require change of the logic, and to provide consistent UX, we should probably to the same in other places where sequence-ish input is accepted.
- There is concern of ambiguity (`str` is a `Sequence[str]`), which causes my irrational feat that it can break in some hard to contain ways (most likely, I am overthinking it).

**Pull request closed without merging**